src/plugins/formulas/formulas.js

```diff
@@ -203,36 +250,231 @@ export class Formulas extends BasePlugin {
-    this.hyperformula.destroy();
-
     super.disablePlugin();
   }
 
+  /**
+   * Triggered on `updateSettings`.
+   */
+  updatePlugin() {
+    this.#settings = this.hot.getSettings()[PLUGIN_KEY];
+
+    this.applyHFSettings();
+
+    if (isDefined(this.#settings.sheetName) && this.#settings.sheetName !== this.sheetName) {
+      this.switchSheet(this.#settings.sheetName);
+    }
+
+    super.updatePlugin();
+  }
+
   /**
    * Destroys the plugin instance.
    */
   destroy() {
+    if (this.hyperformula) {
+      const hfInstances = staticRegister('formulas').getItem('hyperformulaInstances');
+      const sharedHFInstanceUsage = hfInstances ? hfInstances.get(this.hyperformula) : null;
+
+      if (sharedHFInstanceUsage && sharedHFInstanceUsage.includes(this.hot.guid)) {
+        sharedHFInstanceUsage.splice(
+          sharedHFInstanceUsage.indexOf(this.hot.guid),
+          1
+        );
+
+        if (sharedHFInstanceUsage.length === 0) {
+          hfInstances.delete(this.hyperformula);
+          this.hyperformula.destroy();
+        }
+      }
+    }
+
     super.destroy();
   }
 
+  /**
+   * Add a sheet to the shared HyperFormula instance.
+   *
+   * @param {Array} sheetData Data passed to the shared HyperFormula instance. Has to be declared as an array of
+   * arrays - array of objects is not supported in this scenario.
+   * @param {string} [sheetName] The new sheet name. If not provided, will be auto-generated by HyperFormula.
+   * @param {boolean} [autoLoad] If `true`, the new sheet will be loaded into the Handsontable instance.
+   * @returns {boolean} `false` if the data format is unusable, `true` otherwise.
+   */
+  addSheet(sheetData, sheetName, autoLoad) {
+    if (
+      !sheetData ||
+      !Array.isArray(sheetData) ||
+      (sheetData.length && !Array.isArray(sheetData[0]))
+    ) {
+      error('The provided data should be an array of arrays.');
+      return false;
+    }
+
+    const actualSheetName = this.hyperformula.addSheet(sheetName ?? void 0);
+    this.hyperformula.setSheetContent(actualSheetName, sheetData);
+
+    if (autoLoad) {
+      this.switchSheet(actualSheetName);
+    }
+
+    return true;
+  }
+
+  /**
+   * Switch the sheet used as data in the Handsontable instance (it loads the data from the shared HyperFormula
+   * instance).
+   *
+   * @param {string} sheetName Sheet name used in the shared HyperFormula instance.
+   */
+  switchSheet(sheetName) {
+    this.sheetName = sheetName;
+    this.sheetId = this.hyperformula.getSheetId(this.sheetName);
+
+    this.#internal = true;
+    this.hot.loadData(this.hyperformula.getSheetSerialized(this.sheetId));
+    this.#internal = false;
+  }
+
+  /**
+   * Setup the HyperFormula instance. It either creates a new (possibly shared) HyperFormula instance, or attaches
+   * the plugin to an already-existing instance.
+   *
+   * @private
+   * @returns {boolean} `true` if HyperFormula was set up properly, `false` otherwise.
+   */
+  setupHF() {
```

Comment:
It's my functional programmer in me nitpicking, but could this method (and `applyHFSettings` maybe) be refactored not to use so much `this` inside of its body?

If you count the raw lines of code `enablePlugin` did get shorter, but it's still almost logically equivalent - you still have to follow everything from start to finish in your head to understand what's going on, it's just a hundred lines lower. Pretty much the same effect as if you had put the entire body in the place where it's currently being called from. One idea would be to actually use the method parameters, or extract some parts of it into pure functions.

Sorry for being ambiguous, I don't have any concrete code to show as an example, but I hope what I said is relatively clear.

Reply:
I've refactored this entire thing here: 4ecb764. Could you take a look and tell me how you feel about it?

Reply:
I like it! Only one comment https://github.com/handsontable/handsontable/commit/4ecb764c0cf93958ae478927e1637861d56f5e30#r49897867

Reply:
done 91dca03